File: DemoApp/DemoApp/SettingsViewController.m
//
//  SettingsViewController.m
//  DemoApp
//
//  Created by Shimi Sheetrit on 2/3/16.
//  Copyright © 2016 Matomy Media Group Ltd. All rights reserved.
//

#import "SettingsViewController.h"
#import "MainViewController.h"
#import "AppDelegate.h"

#define MOBFOX_HASH_INTER @"267d72ac3f77a3f447b32cf7ebf20673"

@interface SettingsViewController ()

@property (weak, nonatomic) IBOutlet UITextField *hashTextField;
@property (weak, nonatomic) IBOutlet UINavigationBar *navigationBar;

@property (weak, nonatomic) IBOutlet UIView *viewPreview;
@property (weak, nonatomic) IBOutlet UILabel *statusLabel;
@property (weak, nonatomic) IBOutlet UIButton *startButton;
@property (weak, nonatomic) IBOutlet UIBarButtonItem *doneButtonItem;

@property (nonatomic) BOOL isScannerReading;
@property (strong, nonatomic) AVCaptureSession *captureSession;
@property (strong, nonatomic) AVCaptureVideoPreviewLayer *videoPreviewLayer;
@property (strong, nonatomic) MobFoxInterstitialAd *mobfoxInterAd;


@end

@implementation SettingsViewController



- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.isScannerReading = NO;
    self.captureSession = nil;
    
    self.startButton.layer.cornerRadius = 4.0;
    [self.statusLabel setText:@"QR Code Reader is not yet running..."];
    
    [self.view addGestureRecognizer:[[UITapGestureRecognizer alloc]
                                     initWithTarget:self
                                     action:@selector(dismissKeyboard)]];
    
 

}

-(void)dismissKeyboard {
    [_hashTextField resignFirstResponder];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Navigation

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    
    if ([[segue identifier] isEqualToString:@"SettingsToMain"]) {
        
        if (self.hashTextField.text.length > 0) {
            
            MainViewController *vc = [segue destinationViewController];
            vc.invh = self.hashTextField.text;
        }
    }
    
}

- (IBAction)startStopReading:(id)sender {
    
    NSLog(@"startStopReading");
    
    if (!self.isScannerReading) {
        if ([self startReading]) {
            [self.startButton setTitle:@"Stop" forState:UIControlStateNormal];
            [self.statusLabel setText:@"Scanning for QR Code..."];
        }
    }
    else{
        [self stopReading];
        [self.startButton setTitle:@"Start!" forState:UIControlStateNormal];
        [self.statusLabel setText:@"QR Code Reader is not running..."];
    }
    
    self.isScannerReading = !self.isScannerReading;
    
}

- (BOOL)startReading {
    
    NSError *error;
    
    AVCaptureDevice *captureDevice = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    
    AVCaptureDeviceInput *input = [AVCaptureDeviceInput deviceInputWithDevice:captureDevice error:&error];
    if (!input) {
        NSLog(@"%@", [error localizedDescription]);
        return NO;
    }
    
    self.captureSession = [[AVCaptureSession alloc] init];
    [self.captureSession addInput:input];
    
    AVCaptureMetadataOutput *captureMetadataOutput = [[AVCaptureMetadataOutput alloc] init];
    [self.captureSession addOutput:captureMetadataOutput];
    
    dispatch_queue_t dispatchQueue;
    dispatchQueue = dispatch_queue_create("myQueue", NULL);
    [captureMetadataOutput setMetadataObjectsDelegate:self queue:dispatchQueue];
    [captureMetadataOutput setMetadataObjectTypes:[NSArray arrayWithObject:AVMetadataObjectTypeQRCode]];
    
    self.videoPreviewLayer = [[AVCaptureVideoPreviewLayer alloc] initWithSession:self.captureSession];
    [self.videoPreviewLayer setVideoGravity:AVLayerVideoGravityResizeAspectFill];
    [self.videoPreviewLayer setFrame:self.viewPreview.layer.bounds];
    [self.viewPreview.layer addSublayer:self.videoPreviewLayer];
    
    [self.captureSession startRunning];
    
    return YES;
}

-(void)stopReading{
    
    [self.captureSession stopRunning];
    self.captureSession = nil;
    [self.videoPreviewLayer removeFromSuperlayer];
}

#pragma mark AVCaptureMetadataOutputObjectsDelegate

-(void)captureOutput:(AVCaptureOutput *)captureOutput didOutputMetadataObjects:(NSArray *)metadataObjects fromConnection:(AVCaptureConnection *)connection{
    
    if (metadataObjects != nil && [metadataObjects count] > 0) {
        
        AVMetadataMachineReadableCodeObject *metadataObj = [metadataObjects objectAtIndex:0];
        
        if ([[metadataObj type] isEqualToString:AVMetadataObjectTypeQRCode]) {
            
            [_statusLabel performSelectorOnMainThread:@selector(setText:) withObject:@"QR Code Reader is not running..." waitUntilDone:NO];
                        
            [self performSelectorOnMainThread:@selector(stopReading) withObject:nil waitUntilDone:NO];
            
            dispatch_async(dispatch_get_main_queue(), ^{
                
                [self.startButton setTitle:@"Start!" forState:UIControlStateNormal];
                self.hashTextField.text = [metadataObj stringValue];

            });

            self.isScannerReading = NO;
        }
    }
}

#pragma mark MobFox Interstitial Ad Delegate

//best to show after delegate informs an ad was loaded
- (void)MobFoxInterstitialAdDidLoad:(MobFoxInterstitialAd *)interstitial {
    
    NSLog(@"MobFoxInterstitialAdDidLoad:");
    
    if(self.mobfoxInterAd.ready){
        [self.mobfoxInterAd show];
    }
}


@end

File: DemoAppSwift/DemoAppSwift/SettingsViewController.swift
//
//  SettingsViewController.swift
//  DemoAppSwift
//
//  Created by Shimi Sheetrit on 2/16/16.
//  Copyright © 2016 Matomy Media Group Ltd. All rights reserved.
//

import UIKit
import AVFoundation

class SettingsViewController: UIViewController, AVCaptureMetadataOutputObjectsDelegate {
    
    
    private var isScannerReading: Bool!
    private var captureSession: AVCaptureSession!
    private var videoPreviewLayer: AVCaptureVideoPreviewLayer!

    @IBOutlet weak var hashTextField: UITextField!
    @IBOutlet weak var navigationBar: UINavigationBar!
    @IBOutlet weak var viewPreview: UIView!
    @IBOutlet weak var statusLabel: UILabel!
    @IBOutlet weak var startButton: UIButton!
    @IBOutlet weak var doneButtonItem: UIBarButtonItem!
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.isScannerReading = false
        self.captureSession = nil
        
        self.startButton.layer.cornerRadius = 4.0
        self.statusLabel.text = "QR Code Reader is not yet running..."
        
        self.view.addGestureRecognizer(UITapGestureRecognizer.init(target: self, action: #selector(SettingsViewController.dismissKeyboard)))
        
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func dismissKeyboard() {
        
        self.hashTextField.resignFirstResponder()
    }
    
    //MARK: Navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        
        if(segue.identifier == "SettingsToMainSegue") {
            
            if((self.hashTextField.text?.characters.count)! > 0) {
                
                let vc = segue.destination as! MainViewController
                vc.invh = self.hashTextField.text!
                
            }
        }
    }
    
    @IBAction func startStopReading(_ sender: AnyObject) {
        
        if(!self.isScannerReading) {
            if(self.startReading()) {
                
                self.startButton.setTitle("Stop", for:UIControlState())
                self.statusLabel.text = "Scanning for QR Code..."
                
            }
        } else{
            
            self.stopReading()
            self.startButton.setTitle("Start!", for:UIControlState())
            self.statusLabel.text = "QR Code Reader is not running..."
            
        }
        
        self.isScannerReading = !self.isScannerReading
        
    }
    
    func startReading () -> Bool {
        
        var input: AVCaptureDeviceInput
        
        let captureDevice = AVCaptureDevice.defaultDevice(withMediaType: AVMediaTypeVideo)
        
        do {
            
            input = try AVCaptureDeviceInput.init(device: captureDevice!)
            
        }
        catch {
            
            return false
        }
        
        self.captureSession = AVCaptureSession.init()
        self.captureSession.addInput(input)
        
        let captureMetadataOutput = AVCaptureMetadataOutput.init()
        self.captureSession.addOutput(captureMetadataOutput)
        
        var dispatchQueue: DispatchQueue
        dispatchQueue = DispatchQueue(label: "myQueue")
        captureMetadataOutput.setMetadataObjectsDelegate(self, queue: dispatchQueue)
        captureMetadataOutput.metadataObjectTypes = NSArray.init(object: AVMetadataObjectTypeQRCode) as [AnyObject]
        
        self.videoPreviewLayer = AVCaptureVideoPreviewLayer.init(session: self.captureSession)
        self.videoPreviewLayer.videoGravity = AVLayerVideoGravityResizeAspectFill
        self.videoPreviewLayer.frame = self.viewPreview.layer.bounds
        self.viewPreview.layer.addSublayer(self.videoPreviewLayer)
        
        self.captureSession.startRunning()
        
        return true
    }
    
    func stopReading() {
        
        self.captureSession.stopRunning()
        self.captureSession = nil
        self.videoPreviewLayer.removeFromSuperlayer()
    }
    
    //MARK: AVCaptureMetadataOutputObjectsDelegate
    
    private func captureOutput(_ captureOutput: AVCaptureOutput!,
        didOutputMetadataObjects metadataObjects: [AnyObject]!,
        from connection: AVCaptureConnection!) {
            
            if(metadataObjects != nil && metadataObjects.count > 0) {
                
                let metadataObj = metadataObjects[0] as! AVMetadataMachineReadableCodeObject
                
                if(metadataObj.type == AVMetadataObjectTypeQRCode) {
                    
                    //self.statusLabel.performSelector(onMainThread: #selector(setter: UITableViewCell.textLabel), with: "QR Code Reader is not running...", waitUntilDone: false)
                    
                    self.statusLabel.text = "QR Code Reader is not running..."
                    self.performSelector(onMainThread: #selector(SettingsViewController.stopReading), with: nil, waitUntilDone: false)
                    
                    DispatchQueue.main.async(execute: {
               
                        self.startButton.setTitle("Start!", for:UIControlState())
                        self.hashTextField.text = metadataObj.stringValue
                        
                        })
                    
                        self.isScannerReading = false
                    
                }
    
            }
    }
    
    
}
